test/plugin/derivation_endpoint_test.rb

```diff
@@ -0,0 +1,1113 @@
+      response = app.post(derivation_url)
+      assert_equal 405,                              response.status
+      assert_equal "Method not allowed",             response.body_binary
+      assert_equal response.body_binary.length.to_s, response.headers["Content-Length"]
+    end
+  end
+
+  describe "Shrine.derivation_response" do
+    it "works in the main app" do
+      @shrine.plugin :derivation_endpoint, prefix: "derivations"
+
+      derivation_url = @uploaded_file.derivation_url(:gray)
+
+      env = {
+        "REQUEST_METHOD" => "GET",
+        "SCRIPT_NAME"    => "",
+        "PATH_INFO"      => URI(derivation_url).path,
+        "QUERY_STRING"   => URI(derivation_url).query,
+        "rack.input"     => StringIO.new,
+      }
+
+      status, headers, body = @shrine.derivation_response(env)
+
+      assert_equal 200,            status
+      assert_equal "12",           headers["Content-Length"]
+      assert_equal "gray content", body.enum_for(:each).to_a.join
+
+      assert_equal "",                       env["SCRIPT_NAME"]
+      assert_equal URI(derivation_url).path, env["PATH_INFO"]
+    end
+
+    it "works in a mounted app" do
+      @shrine.plugin :derivation_endpoint, prefix: "derivations"
+
+      derivation_url = @uploaded_file.derivation_url(:gray)
+
+      env = {
+        "REQUEST_METHOD" => "GET",
+        "SCRIPT_NAME"    => "/foo",
+        "PATH_INFO"      => URI(derivation_url).path,
+        "QUERY_STRING"   => URI(derivation_url).query,
+        "rack.input"     => StringIO.new,
+      }
+
+      status, headers, body = @shrine.derivation_response(env)
+
+      assert_equal 200,    status
+      assert_equal "/foo", env["SCRIPT_NAME"]
+    end
+
+    it "accepts additional options" do
+      @shrine.plugin :derivation_endpoint, prefix: "derivations"
+
+      derivation_url = @uploaded_file.derivation_url(:gray)
+
+      env = {
+        "REQUEST_METHOD" => "GET",
+        "SCRIPT_NAME"    => "",
+        "PATH_INFO"      => URI(derivation_url).path,
+        "QUERY_STRING"   => URI(derivation_url).query,
+        "rack.input"     => StringIO.new,
+      }
+
+      status, headers, body = @shrine.derivation_response(env, type: "text/plain")
+
+      assert_equal 200,          status
+      assert_equal "text/plain", headers["Content-Type"]
+    end
+
+    it "fails when request path doesn't start with prefix" do
+      @shrine.plugin :derivation_endpoint, prefix: "derivations"
+
+      derivation_url = @uploaded_file.derivation_url(:gray, prefix: "other")
+
+      env = {
+        "REQUEST_METHOD" => "GET",
+        "SCRIPT_NAME"    => "",
+        "PATH_INFO"      => URI(derivation_url).path,
+        "QUERY_STRING"   => URI(derivation_url).query,
+        "rack.input"     => StringIO.new,
+      }
+
+      assert_raises(Shrine::Error) do
+        @shrine.derivation_response(env)
+      end
+    end
+  end
+
+  describe "UploadedFile#derivation_response" do
+    describe "without upload" do
```

Comment:
These are referring to the 3 modes: always_process, cache_processed, and redirect_to_cache_processed